// @primer/react/lib/Token/AvatarToken.js
'use strict';

var React = require('react');
var styled = require('styled-components');
var constants = require('../constants.js');
var TokenBase = require('./TokenBase.js');
var Token = require('./Token.js');
var Avatar = require('../Avatar/Avatar.js');
var toggleStyledComponent = require('../internal/utils/toggleStyledComponent.js');
require('../FeatureFlags/FeatureFlags.js');
var useFeatureFlag = require('../FeatureFlags/useFeatureFlag.js');
require('../FeatureFlags/DefaultFeatureFlags.js');
var AvatarToken_module = require('./AvatarToken.module.css.js');
var clsx = require('clsx');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var styled__default = /*#__PURE__*/_interopDefault(styled);

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }

// TODO: update props to only accept 'large' and 'xlarge' on the next breaking change

const CSS_MODULES_FEATURE_FLAG = 'primer_react_css_modules_staff';
const AvatarContainer = toggleStyledComponent.toggleStyledComponent(CSS_MODULES_FEATURE_FLAG, 'span', styled__default.default.span.withConfig({
  displayName: "AvatarToken__AvatarContainer",
  componentId: "sc-62sih1-0"
})(["--spacing:calc(", " * 2);display:block;height:", ";width:", ";"], constants.get('space.1'), props => `calc(${TokenBase.tokenSizes[props.avatarSize]} - var(--spacing))`, props => `calc(${TokenBase.tokenSizes[props.avatarSize]} - var(--spacing))`));
const AvatarToken = /*#__PURE__*/React.forwardRef(({
  avatarSrc,
  id,
  size = TokenBase.defaultTokenSize,
  className,
  ...rest
}, forwardedRef) => {
  const enabled = useFeatureFlag.useFeatureFlag(CSS_MODULES_FEATURE_FLAG);
  if (enabled) {
    return /*#__PURE__*/React__default.default.createElement(Token, _extends({
      leadingVisual: () => /*#__PURE__*/React__default.default.createElement(AvatarContainer, {
        avatarSize: size,
        className: AvatarToken_module.AvatarContainer,
        "data-size": size
      }, /*#__PURE__*/React__default.default.createElement(Avatar.default, {
        src: avatarSrc,
        size: parseInt(TokenBase.tokenSizes[size], 10),
        className: AvatarToken_module.Avatar
      })),
      size: size,
      id: id === null || id === undefined ? undefined : id.toString(),
      className: clsx.clsx(AvatarToken_module.Token, className)
    }, rest, {
      ref: forwardedRef
    }));
  }
  return /*#__PURE__*/React__default.default.createElement(Token, _extends({
    leadingVisual: () => /*#__PURE__*/React__default.default.createElement(AvatarContainer, {
      avatarSize: size
    }, /*#__PURE__*/React__default.default.createElement(Avatar.default, {
      src: avatarSrc,
      size: parseInt(TokenBase.tokenSizes[size], 10),
      sx: {
        width: '100%',
        height: '100%'
      }
    })),
    size: size,
    id: id === null || id === undefined ? undefined : id.toString(),
    sx: {
      paddingLeft: constants.get('space.1')
    },
    className: className
  }, rest, {
    ref: forwardedRef
  }));
});
AvatarToken.displayName = 'AvatarToken';

module.exports = AvatarToken;
