// @primer/react/lib/internal/components/TextInputInnerAction.js
'use strict';

var React = require('react');
var Box = require('../../Box/Box.js');
var IconButton = require('../../Button/IconButton.js');
require('../../Button/ButtonBase.js');
require('../../utils/defaultSxProp.js');
var Button = require('../../Button/Button.js');
var Tooltip = require('../../TooltipV2/Tooltip.js');
require('@styled-system/css');
var merge = require('deepmerge');
var clsx = require('clsx');
var TextInputInnerAction_module = require('./TextInputInnerAction.module.css.js');
require('../../FeatureFlags/FeatureFlags.js');
var useFeatureFlag = require('../../FeatureFlags/useFeatureFlag.js');
require('../../FeatureFlags/DefaultFeatureFlags.js');
var UnstyledTextInput = require('./UnstyledTextInput.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);
var merge__default = /*#__PURE__*/_interopDefault(merge);

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const invisibleButtonStyleOverrides = {
  paddingTop: '2px',
  paddingRight: '4px',
  paddingBottom: '2px',
  paddingLeft: '4px',
  position: 'relative',
  backgroundColor: 'transparent',
  color: 'fg.subtle',
  '&:hover, &:focus': {
    color: 'fg.default'
  },
  '&[data-component="IconButton"]': {
    width: 'var(--inner-action-size)',
    height: 'var(--inner-action-size)'
  },
  '@media (pointer: coarse)': {
    ':after': {
      content: '""',
      position: 'absolute',
      left: 0,
      right: 0,
      transform: 'translateY(-50%)',
      top: '50%',
      minHeight: '44px'
    }
  }
};
const ConditionalTooltip = ({
  'aria-label': ariaLabel,
  children,
  tooltipDirection
}) => /*#__PURE__*/React__default.default.createElement(React__default.default.Fragment, null, ariaLabel ? /*#__PURE__*/React__default.default.createElement(Tooltip.Tooltip, {
  text: ariaLabel,
  direction: tooltipDirection,
  sx: {
    /* inline-block is used to ensure the tooltip dimensions don't
       collapse when being used with `grid` or `inline` children */
    display: 'inline-block'
  }
}, children) : children);
const TextInputAction = /*#__PURE__*/React.forwardRef(({
  'aria-label': ariaLabel,
  'aria-labelledby': ariaLabelledBy,
  tooltipDirection,
  children,
  icon,
  sx: sxProp,
  className,
  variant = 'invisible',
  ...rest
}, forwardedRef) => {
  const enabled = useFeatureFlag.useFeatureFlag(UnstyledTextInput.TEXT_INPUT_CSS_MODULES_FEATURE_FLAG);
  const styleProps = enabled ? {
    className: clsx.clsx(variant === 'invisible' && TextInputInnerAction_module.Invisible, className),
    sx: sxProp || {}
  } : {
    className,
    sx: variant === 'invisible' ? merge__default.default(invisibleButtonStyleOverrides, sxProp || {}) : sxProp || {}
  };
  if (icon && !ariaLabel || !children && !ariaLabel) {
    // eslint-disable-next-line no-console
    console.warn('Use the `aria-label` prop to provide an accessible label for assistive technology');
  }
  const accessibleLabel = ariaLabel ? {
    'aria-label': ariaLabel
  } : ariaLabelledBy ? {
    'aria-labelledby': ariaLabelledBy
  } : {
    'aria-label': ''
  };
  return /*#__PURE__*/React__default.default.createElement(Box, {
    as: "span",
    className: "TextInput-action",
    marginLeft: 1,
    marginRight: 1,
    lineHeight: "0"
  }, icon && !children && ariaLabel ? /*#__PURE__*/React__default.default.createElement(IconButton.IconButton, _extends({}, accessibleLabel, {
    tooltipDirection: tooltipDirection !== null && tooltipDirection !== undefined ? tooltipDirection : 's',
    variant: variant,
    type: "button",
    icon: icon,
    size: "small"
  }, styleProps, rest, {
    ref: forwardedRef
  })) : /*#__PURE__*/React__default.default.createElement(ConditionalTooltip, {
    "aria-label": ariaLabel
  }, /*#__PURE__*/React__default.default.createElement(Button.ButtonComponent, _extends({
    variant: variant,
    type: "button"
  }, styleProps, rest, {
    ref: forwardedRef
  }), children)));
});

module.exports = TextInputAction;
