// @primer/react/lib/internal/components/UnstyledTextInput.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styled = require('styled-components');
var sx = require('../../sx.js');
var toggleStyledComponent = require('../utils/toggleStyledComponent.js');
var React = require('react');
require('../../FeatureFlags/FeatureFlags.js');
var useFeatureFlag = require('../../FeatureFlags/useFeatureFlag.js');
require('../../FeatureFlags/DefaultFeatureFlags.js');
var UnstyledTextInput_module = require('./UnstyledTextInput.module.css.js');
var clsx = require('clsx');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var styled__default = /*#__PURE__*/_interopDefault(styled);
var React__default = /*#__PURE__*/_interopDefault(React);

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const TEXT_INPUT_CSS_MODULES_FEATURE_FLAG = 'primer_react_css_modules_ga';
const ToggledUnstyledTextInput = toggleStyledComponent.toggleStyledComponent(TEXT_INPUT_CSS_MODULES_FEATURE_FLAG, 'input', styled__default.default.input.withConfig({
  displayName: "UnstyledTextInput__ToggledUnstyledTextInput",
  componentId: "sc-14ypya-0"
})(["border:0;font-size:inherit;font-family:inherit;background-color:transparent;-webkit-appearance:none;color:inherit;width:100%;&:focus{outline:0;}", ";"], sx.default));
const UnstyledTextInput = /*#__PURE__*/React__default.default.forwardRef(function UnstyledTextInput({
  className,
  ...rest
}, forwardRef) {
  const enabled = useFeatureFlag.useFeatureFlag(TEXT_INPUT_CSS_MODULES_FEATURE_FLAG);
  return /*#__PURE__*/React__default.default.createElement(ToggledUnstyledTextInput, _extends({
    ref: forwardRef
  }, rest, {
    className: clsx.clsx(className, enabled && UnstyledTextInput_module.Input)
  }));
});
UnstyledTextInput.displayName = 'UnstyledTextInput';

exports.TEXT_INPUT_CSS_MODULES_FEATURE_FLAG = TEXT_INPUT_CSS_MODULES_FEATURE_FLAG;
exports.default = UnstyledTextInput;
