// @primer/react/lib/Header/Header.js
'use strict';

var React = require('react');
var clsx = require('clsx');
var Header_module = require('./Header.module.css.js');
var defaultSxProp = require('../utils/defaultSxProp.js');
var Box = require('../Box/Box.js');

function _interopDefault (e) { return e && e.__esModule ? e : { default: e }; }

var React__default = /*#__PURE__*/_interopDefault(React);

function _extends() { return _extends = Object.assign ? Object.assign.bind() : function (n) { for (var e = 1; e < arguments.length; e++) { var t = arguments[e]; for (var r in t) ({}).hasOwnProperty.call(t, r) && (n[r] = t[r]); } return n; }, _extends.apply(null, arguments); }
const Header = /*#__PURE__*/React__default.default.forwardRef(function Header({
  children,
  className,
  sx: sxProp = defaultSxProp.defaultSxProp,
  as = 'header',
  ...rest
}, forwardRef) {
  if (sxProp !== defaultSxProp.defaultSxProp || as !== 'header') {
    return /*#__PURE__*/React__default.default.createElement(Box, _extends({
      as: as,
      sx: sxProp,
      ref: forwardRef,
      className: clsx.clsx(className, Header_module.Header)
    }, rest), children);
  }
  return /*#__PURE__*/React__default.default.createElement("header", _extends({
    ref: forwardRef,
    className: clsx.clsx(className, Header_module.Header)
  }, rest), children);
});
Header.displayName = 'Header';
const HeaderItem = /*#__PURE__*/React__default.default.forwardRef(function HeaderItem({
  children,
  className,
  sx: sxProp = defaultSxProp.defaultSxProp,
  full,
  ...rest
}, forwardRef) {
  if (sxProp !== defaultSxProp.defaultSxProp) {
    return /*#__PURE__*/React__default.default.createElement(Box, _extends({
      as: 'div',
      sx: sxProp,
      ref: forwardRef,
      className: clsx.clsx(className, Header_module.HeaderItem),
      "data-full": full
    }, rest), children);
  }
  return /*#__PURE__*/React__default.default.createElement("div", _extends({
    ref: forwardRef,
    className: clsx.clsx(className, Header_module.HeaderItem),
    "data-full": full
  }, rest), children);
});
HeaderItem.displayName = 'Header.Item';
const HeaderLink = /*#__PURE__*/React__default.default.forwardRef(function HeaderLink({
  children,
  className,
  sx: sxProp = defaultSxProp.defaultSxProp,
  as = 'a',
  ...rest
}, forwardRef) {
  if (sxProp !== defaultSxProp.defaultSxProp || as !== 'a') {
    return /*#__PURE__*/React__default.default.createElement(Box, _extends({
      as: as,
      sx: sxProp,
      ref: forwardRef,
      className: clsx.clsx(className, Header_module.HeaderLink)
    }, rest), children);
  }
  return /*#__PURE__*/React__default.default.createElement("a", _extends({
    ref: forwardRef,
    className: clsx.clsx(className, Header_module.HeaderLink)
  }, rest), children);
});
HeaderLink.displayName = 'Header.Link';
var Header$1 = Object.assign(Header, {
  Link: HeaderLink,
  Item: HeaderItem
});

module.exports = Header$1;
